File: programmers/Dynamic programming/dp-03.py
#
# https://programmers.co.kr/learn/courses/30/lessons/42898?language=cpp
# level : 3 / 등굣길
#

def solution(m, n, puddles):
    answer, mod = 0, 1e9+7
    map = [[1 for _ in range(m)] for _ in range(n)]
 
    for puddle in puddles:
        map[puddle[1]-1][puddle[0]-1] = 0
      
    for i in range(1, m-1):
        if map[0][i] == 0:
            map[0][i+1] = 0

    for i in range(1, n-1):
        if map[i][0] == 0:
            map[i+1][0] = 0
            
    
    for i in range(1,n):
        for j in range(1, m):
            if map[i][j] == 0:
                continue
            map[i][j] = int((map[i-1][j] + map[i][j-1]) % mod)
 
    return map[n-1][m-1]

File: programmers/Dynamic programming/dp-03.cpp
/* 
 *
 * 
 */

#include <string>
#include <vector>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    int answer = 0, mod = 1e9+7;
    vector<vector<int>> v(n, vector<int>(m, 1));
    
    for(auto puddle : puddles) v[puddle[1]-1][puddle[0]-1] = 0;

    for(int i = 1; i < m; i++){
        if(v[0][i] == 0){
            for(int j = i; j < m; j++) v[0][j] = 0;
            break;
        }
    }
    for(int i = 1; i < n; i++){
        if(v[i][0] == 0){
            for(int j = i; j < n; j++) v[j][0] = 0;
            break;
        }
    }
    
    for(int i = 1; i < n; i++){
        for(int j = 1; j < m; j++){
            if(v[i][j] == 0) continue;
            v[i][j] = (v[i-1][j] + v[i][j-1]) % mod;
        }
    }
    return v[n-1][m-1];
}